components/brave_wallet/browser/brave_wallet_p3a.cc

```diff
@@ -0,0 +1,83 @@
+/* Copyright 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <string>
+
+#include "base/metrics/histogram_macros.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_p3a.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_service.h"
+#include "brave/components/brave_wallet/browser/keyring_service.h"
+#include "brave/components/brave_wallet/browser/pref_names.h"
+#include "components/prefs/pref_service.h"
+
+namespace brave_wallet {
+
+// Has the Wallet keyring been created?
+// i) No, ii) Yes
+void RecordKeyringCreated(mojom::KeyringInfoPtr keyring_info) {
+  const int max_bucket = 2;
+  UMA_HISTOGRAM_EXACT_LINEAR(
+      "Brave.Wallet.KeyringCreated",
+      static_cast<int>(keyring_info->is_default_keyring_created), max_bucket);
+}
+
+// What is the DefaultWalletSetting?
+// i) AskDeprecated, ii) None, ii) CryptoWallets,
+// iv) BraveWalletPreferExtension, v) BraveWallet
+void RecordDefaultWalletSetting(PrefService* pref_service) {
+  const int max_bucket = 5;
+  auto default_wallet = pref_service->GetInteger(kDefaultWallet2);
+  UMA_HISTOGRAM_EXACT_LINEAR("Brave.Wallet.DefaultWalletSetting",
+                             default_wallet, max_bucket);
+}
+
+BraveWalletP3A::BraveWalletP3A(BraveWalletService* wallet_service,
+                               KeyringService* keyring_service,
+                               PrefService* pref_service)
+    : wallet_service_(wallet_service),
+      keyring_service_(keyring_service),
+      pref_service_(pref_service) {
+  RecordInitialBraveWalletP3AState();
+  wallet_service_->AddObserver(
+      wallet_service_observer_receiver_.BindNewPipeAndPassRemote());
+  keyring_service_->AddObserver(
+      keyring_service_observer_receiver_.BindNewPipeAndPassRemote());
+}
+
+BraveWalletP3A::~BraveWalletP3A() = default;
+
+void BraveWalletP3A::RecordInitialBraveWalletP3AState() {
+  keyring_service_->GetKeyringInfo(mojom::kDefaultKeyringId,
+                                   base::BindOnce(&RecordKeyringCreated));
+  RecordDefaultWalletSetting(pref_service_);
+}
+
+// KeyringServiceObserver
+void BraveWalletP3A::KeyringCreated() {
+  keyring_service_->GetKeyringInfo(mojom::kDefaultKeyringId,
+                                   base::BindOnce(&RecordKeyringCreated));
+}
+void BraveWalletP3A::KeyringRestored() {}
```

Comment:
couldn't you put these to `.h`?

Reply:
Because you're asking, I'd guess that I probably should.  I do see some examples of empty implementations in .cc files for constructors and destructors, e.g. https://github.com/brave/brave-core/blob/b6766f7eab79a79c6a4e1e18a9d1feeffbd8f824/components/brave_wallet/common/eth_address.cc#L24

Is the convention to define empty functions in header files unless they are constructors / destructors?

Reply:
Updated in https://github.com/brave/brave-core/pull/11929/commits/3432f0ad7540df051620f5d3c130386352502d4f

Reply:
This is just handy @nvonpentz , saves some scrolling in .cc